pkg/reconciler/cache/cachedresources/replication/replication_controller.go

```diff
@@ -0,0 +1,240 @@
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+	gvrKey := fmt.Sprintf("%s.%s.%s::%s", gvr.Version, gvr.Resource, gvr.Group, key)
+	c.queue.Add(gvrKey)
+}
+
+func (c *Controller) enqueueCacheObject(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	// This way we extract what is the original GVR of the object that we are replicating.
+	cr, ok := obj.(*cachev1alpha1.CachedObject)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expected *cachev1alpha1.CachedObject, got %T", obj))
+		return
+	}
+
+	labels := cr.GetLabels()
+	gvr := schema.GroupVersionResource{
+		Group:    labels[LabelKeyObjectGroup],
+		Version:  labels[LabelKeyObjectVersion],
+		Resource: labels[LabelKeyObjectResource],
+	}
+
+	gvrKey := fmt.Sprintf("%s.%s.%s::%s", gvr.Version, gvr.Resource, gvr.Group, key)
+	c.queue.Add(gvrKey)
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *Controller) Start(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+	defer func() {
+		for _, closure := range c.closures {
+			closure()
+		}
+	}()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(cacheclient.WithShardInContext(ctx, shard.New(c.shardName)), logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+	c.started = true
+	<-ctx.Done()
+	c.started = false
+}
+
+func (c *Controller) Started() bool {
+	return c.started
+}
+
+func (c *Controller) SetLabelSelector(localLabelSelector labels.Selector) {
+	c.localLabelSelector = localLabelSelector
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	gvrKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(gvrKey)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), gvrKey)
+	ctx = klog.NewContext(ctx, logger)
+	err := c.reconcile(ctx, gvrKey)
+	if err == nil {
+		c.queue.Forget(gvrKey)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with: %w", gvrKey, err))
+	c.queue.AddRateLimited(gvrKey)
+
+	return true
+}
+
+func (c *Controller) SetDeleted(ctx context.Context) {
+	c.deleted = true
+}
+
+type Controller struct {
+	shardName string
+	queue     workqueue.TypedRateLimitingInterface[string]
+
+	dynamicCacheClient kcpdynamic.ClusterInterface
+	kcpCacheClient     kcpclientset.ClusterInterface
+
+	replicated *ReplicatedGVR
+	gvr        schema.GroupVersionResource
+
+	// callback is called when we want to trigger parent object reconciliation
+	callback func()
+	// closures are cleanup functions that are called when the controller is stopped.
+	closures []func()
```

Comment:
The general term closure usually refers to a function with a scope; it's not clear what `closures` does without reading the comment. Just `cleanupFuncs`? Or `cleaners` to not have Hungarian notation...

Reply:
Rename to `cleanupFuncs`.